pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -211,14 +205,14 @@ func (w *Watcher) traversePluginDir(dir string) error {
 				return fmt.Errorf("failed to watch %s, err: %v", path, err)
 			}
 		case mode&os.ModeSocket != 0:
-			w.wg.Add(1)
-			go func() {
-				defer w.wg.Done()
-				w.fsWatcher.Events <- fsnotify.Event{
-					Name: path,
-					Op:   fsnotify.Create,
-				}
-			}()
+			event := fsnotify.Event{
+				Name: path,
+				Op:   fsnotify.Create,
+			}
+			//TODO: Handle errors by taking corrective measures
+			if err := w.handleCreateEvent(event); err != nil {
```

Comment:
Looking at the implementation of handleCreateEvent, I think it's more direct to just do:
```go
if !w.containsBlacklistedDir(path) {
  return w.traversePluginDir(path)
}
```

Reply:
Not sure if I follow, do you mean inside `handleCreateEvent`? If so, then it wouldn't call `handlePluginRegistration`... or am I missing somehting?

Reply:
I mean "handle create event" doesn't really seem like what this is doing, unless I'm not following the control flow. If you look at what handle create event is doing, the snippet I commented above is all that gets hit in that function, so I'm suggesting to just inline that piece here.

Reply:
Ah, OK... I understand now...

This part of the code creates "synthetic create events" if we find a socket file. Then we "handle" this event right here (as opposed to sending it to a processing goroutine like it was before).

Since we're dealing with a socket file, we  don't really want to `traversePluginDir()` it, but instead we want to register the plugin that's listening to this unix domain socket file (i.e., `handlePluginRegistration()`).

However, we also need to make sure that we ignore black-listed directories (like you pointed in the snippet above) and files prefixed with a "."; this would do it:

```go
			if !w.containsBlacklistedDir(path) {
				if !strings.HasPrefix(path, ".") {
					// TODO: Handle errors by taking corrective measures
					if err := w.handlePluginRegistration(path); err != nil {
						klog.Errorf("error %v when handling create event for file: %s", err, path)
					}
				}
			}
```

IMO `handleCreateEvent()` looks like would be a better fit though (since we're almost re-implementing it in-line).

Does this make sense?

Reply:
Ah, sorry, I completely misread what `handleCreateEvent` was doing.